File: Documentation/5.4/Samples/python/ClientApi/Session/ClusterTransaction/AtomicGuards.py
from ravendb import SessionOptions, TransactionMode, GetCompareExchangeValuesOperation
from ravendb.infrastructure.entities import User

from examples_base import ExampleBase


class LoadingEntities(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_create_atomic_guard(self):
        # nodes, leader = create_raft_cluster(3)
        with self.embedded_server.get_document_store("AtomicGuard") as store:
            # region atomic_guards_enabled
            with store.open_session(
                # Open a cluster-wide session:
                session_options=SessionOptions(transaction_mode=TransactionMode.CLUSTER_WIDE)
            ) as session:
                session.store(User(), "users/johndoe")
                session.save_changes()
                # An atomic-guard is now automatically created for the new document "users/johndoe"

            # Open two concurrent cluster-wide sessions:
            with store.open_session(
                session_options=SessionOptions(transaction_mode=TransactionMode.CLUSTER_WIDE)
            ) as session1:
                with store.open_session(
                    session_options=SessionOptions(transaction_mode=TransactionMode.CLUSTER_WIDE)
                ) as session2:
                    # Both sessions load the same document:
                    loaded_user_1 = session1.load("users/johndoe", User)
                    loaded_user_1.name = "jindoe"
                    loaded_user_2 = session2.load("users/johndoe", User)
                    loaded_user_2.name = "jandoe"

                    # session1 saves its changes first —
                    # this increments the Raft index of the associated atomic guard.
                    session1.save_changes()

                    # session2 tries to save using an outdated atomic guard version
                    # and fails with a ConcurrencyException.
                    session2.save_changes()
            # endregion

            result = store.operations.send(GetCompareExchangeValuesOperation("", User))

    def test_do_not_create_atomic_guard(self):
        with self.embedded_server.get_document_store("NotAtomicGuard") as store:
            # region atomic_guards_disabled
            with store.open_session(
                # Open a cluster-wide session
                session_options=SessionOptions(
                    transaction_mode=TransactionMode.CLUSTER_WIDE,
                    disable_atomic_document_writes_in_cluster_wide_transaction=True,
                )
            ) as session:
                session.store(User(), "users/johndoe")

                # No atomic-guard will be created upon save_changes
                session.save_changes()
            # endregion

            result = store.operations.send(GetCompareExchangeValuesOperation("", User))
            
    def load_before_storing(self):
            with self.embedded_server.get_document_store("LoadBeforeStore") as store:
                # region load_before_storing
                with store.open_session(
                    session_options=SessionOptions(
                        # Open a cluster-wide session
                        transaction_mode=TransactionMode.CLUSTER_WIDE
                    )
                ) as session:
                    # Load the user document BEFORE creating or updating
                    user = session.load("users/johndoe", User)
        
                    if user is None:
                        # Document doesn't exist => create a new document
                        new_user = User()
                        new_user.id = "users/johndoe"
                        new_user.name = "John Doe"
                        # ... initialize other properties
        
                        # Store the new user document in the session
                        session.store(new_user)
                    else:
                        # Document exists => apply your modifications
                        user.name = "New name"
                        # ... make any other updates
                        
                        # No need to call store() again
                        # RavenDB tracks changes on loaded entities
        
                    # Commit your changes
                    session.save_changes()
                # endregion

File: Documentation/5.4/Samples/php/ClientApi/Session/ClusterTransaction/AtomicGuards.php
<?php

use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Operations\CompareExchange\GetCompareExchangeValuesOperation;
use RavenDB\Documents\Session\SessionOptions;
use RavenDB\Documents\Session\TransactionMode;
use RavenDB\Samples\Infrastructure\Entity\User;

class AtomicGuards
{
    public function testCreateAtomicGuard(): void
    {
        //var (nodes, leader) = await CreateRaftCluster(3);
            $store = new DocumentStore(
                [ "http://127.0.0.1:8080" ],
                "test"
            );
            $store->initialize();

            # region atomic_guards_enabled
            // Open a cluster-wide session:
            $sessionOptions = new SessionOptions();
            $sessionOptions->setTransactionMode(TransactionMode::clusterWide());

            $session = $store->openSession($sessionOptions);

            try {
                $session->store(new User(), "users/johndoe");
                // An atomic-guard is now automatically created for the new document "users/johndoe".
                $session->saveChanges();
            } finally {
                $session->close();
            }

            // Open two concurrent cluster-wide sessions:

            $sessionOptions1 = new SessionOptions();
            $sessionOptions1->setTransactionMode(TransactionMode::clusterWide());

            $session1 = $store->openSession($sessionOptions1);
            try {
                $sessionOptions2 = new SessionOptions();
                $sessionOptions2->setTransactionMode(TransactionMode::clusterWide());
                $session2 = $store->openSession($sessionOptions2);

                try {
                // Both sessions load the same document:
                var $loadedUser1 = $session1->load(User::class, "users/johndoe");
                $loadedUser1->setName("jindoe");

                $loadedUser2 = $session2->load(User::class, "users/johndoe");
                $loadedUser2->setName("jandoe");

                // session1 saves its changes first —
                // this increments the Raft index of the associated atomic guard.
                $session1->saveChanges();

                // session2 tries to save using an outdated atomic guard version
                // and fails with a ConcurrencyException.
                $session2->saveChanges();
                } finally {
                    $session2->close();
                }

            } finally {
                $session1->close();
            }
            # endregion

            $result = $store->operations()->sendAsync(new GetCompareExchangeValuesOperation(User::class, ""));
    }

    public function testDoNotCreateAtomicGuard(): void
    {
        //var (nodes, leader) = await CreateRaftCluster(3);
        $store = new DocumentStore(
            [ "http://127.0.0.1:8080" ],
            "test"
        );
        $store->initialize();

        # region atomic_guards_disabled
        $sessionOptions = new SessionOptions();
        $sessionOptions->setTransactionMode(TransactionMode::clusterWide());
        $sessionOptions->setDisableAtomicDocumentWritesInClusterWideTransaction(true);

        $session = $store->openSession($sessionOptions);

        try {
            $session->store(new User(), "users/johndoe");
            // No atomic-guard will be created upon saveChanges
            $session->saveChanges();
        } finally {
            $session->close();
        }
        # endregion

        $result = $store->operations()->sendAsync(new GetCompareExchangeValuesOperation(User::class, ""));
    }

    public function loadBeforeStoring(): void
    {
        $store = new DocumentStore(
            [ "http://127.0.0.1:8080" ],
            "test"
        );
        $store->initialize();

        # region load_before_storing
        // Open a cluster-wide session
        $sessionOptions = new SessionOptions();
        $sessionOptions->setTransactionMode(TransactionMode::clusterWide());

        $session = $store->openSession($sessionOptions);
        try {
            // Load the user document BEFORE creating or updating
            $user = $session->load(User::class, "users/johndoe");

            if ($user === null) {
                // Document doesn't exist => create a new document:
                $newUser = new User();
                $newUser->setName("John Doe");
                // ... initialize other properties

                // Store the new user document in the session
                $session->store($newUser, "users/johndoe");
            } else {
                // Document exists => apply your modifications:
                $user->setName("New name");
                // ... make any other updates
                
                // No need to call Store() again
                // RavenDB tracks changes on loaded entities
            }

            // Commit your changes
            $session->saveChanges();
        } finally {
            $session->close();
        }
        # endregion
    }
}
